xtask/src/release.rs

```diff
@@ -71,26 +78,157 @@ impl flags::Release {
+            .run()
+            .context("FAILED to merge new commits, something went wrong")?;
+
+        drop(josh);
+        Ok(())
+    }
+}
+
+impl flags::Push {
+    pub(crate) fn run(self, sh: &Shell) -> anyhow::Result<()> {
+        let branch = "sync-from-ra";
+        let Ok(github_user) = env::var("GITHUB_USER") else {
+            bail!("please set `GITHUB_USER` to the GitHub username of your `rust-lang/rust` clone");
+        };
+
+        sh.change_dir(project_root());
+        let base = sh.read_file("rust-version")?.trim().to_owned();
+        // Make sure the repo is clean.
+        if !cmd!(sh, "git status --untracked-files=no --porcelain").read()?.is_empty() {
+            bail!("working directory must be clean before running `cargo xtask push`");
+        }
+        // Make sure josh is running.
+        let josh = start_josh()?;
+
+        // Find a repo we can do our preparation in.
+        if let Ok(rustc_git) = env::var("RUSTC_GIT") {
+            // If rustc_git is `Some`, we'll use an existing fork for the branch updates.
+            sh.change_dir(rustc_git);
+        } else {
+            bail!("please set `RUSTC_GIT` to a `rust-lang/rust` clone");
+        };
+        // Prepare the branch. Pushing works much better if we use as base exactly
+        // the commit that we pulled from last time, so we use the `rust-version`
+        // file to find out which commit that would be.
+        println!("Preparing {github_user}/rust (base: {base})...");
+        if cmd!(sh, "git fetch https://github.com/{github_user}/rust {branch}")
+            .ignore_stderr()
+            .read()
+            .is_ok()
+        {
+            bail!(
+                "The branch '{branch}' seems to already exist in 'https://github.com/{github_user}/rust'. Please delete it and try again."
+            );
+        }
+        cmd!(sh, "git fetch https://github.com/rust-lang/rust {base}").run()?;
+        cmd!(sh, "git push https://github.com/{github_user}/rust {base}:refs/heads/{branch}")
+            .ignore_stdout()
+            .ignore_stderr() // silence the "create GitHub PR" message
             .run()?;
+        println!();
+
+        // Do the actual push.
+        sh.change_dir(project_root());
+        println!("Pushing rust-analyzer changes...");
+        cmd!(
+            sh,
+            "git push http://localhost:{JOSH_PORT}/{github_user}/rust.git{JOSH_FILTER}.git HEAD:{branch}"
+        )
+        .run()?;
+        println!();
+
+        // Do a round-trip check to make sure the push worked as expected.
+        cmd!(
+            sh,
+            "git fetch http://localhost:{JOSH_PORT}/{github_user}/rust.git{JOSH_FILTER}.git {branch}"
+        )
+        .ignore_stderr()
+        .read()?;
+        let head = cmd!(sh, "git rev-parse HEAD").read()?;
+        let fetch_head = cmd!(sh, "git rev-parse FETCH_HEAD").read()?;
+        if head != fetch_head {
```

Comment:
FWIW this is pretty crucial. If you continue and merge this into rustc despite this check failing, the RA repo may need a force-push to fix the situation. That's what happened in Miri.

I haven't seen this check fail in over a year.

Reply:
I managed to do it:

```
$ git fetch https://github.com/rust-lang/rust 4e431fad67b46c480f1833119cd368fa33df95f7
remote: Enumerating objects: 161, done.
remote: Counting objects: 100% (120/120), done.
remote: Compressing objects: 100% (15/15), done.
remote: Total 161 (delta 107), reused 109 (delta 105), pack-reused 41
Receiving objects: 100% (161/161), 72.28 KiB | 7.23 MiB/s, done.
Resolving deltas: 100% (108/108), completed with 76 local objects.
From https://github.com/rust-lang/rust
 * branch                    4e431fad67b46c480f1833119cd368fa33df95f7 -> FETCH_HEAD
$ git push https://github.com/lnicola/rust 4e431fad67b46c480f1833119cd368fa33df95f7:refs/heads/sync-from-ra

Pushing rust-analyzer changes...
$ git push http://localhost:42042/lnicola/rust.git:rev(f5a9250147f6569d8d89334dc9cca79c0322729f:prefix=src/tools/rust-analyzer):/src/tools/rust-analyzer.git HEAD:sync-from-ra
Enumerating objects: 1335, done.
Counting objects: 100% (851/851), done.
Delta compression using up to 32 threads
Compressing objects: 100% (293/293), done.
Writing objects: 100% (500/500), 97.33 KiB | 48.67 MiB/s, done.
Total 500 (delta 332), reused 351 (delta 194), pack-reused 0 (from 0)
remote: Resolving deltas: 100% (332/332), completed with 143 local objects.
remote: josh-proxy
remote: response from upstream:
remote: To https://github.com/lnicola/rust.git
remote:    4e431fad67b..cc7848da115  JOSH_PUSH -> sync-from-ra
remote: REWRITE(1455ee0766bcb6075e6fdf797d24953f0af52b20 -> bf5c131477627b5e6273d5338d6b3ea137db4d41)
remote: 
remote: 
To http://localhost:42042/lnicola/rust.git:rev(f5a9250147f6569d8d89334dc9cca79c0322729f:prefix=src/tools/rust-analyzer):/src/tools/rust-analyzer.git
   16b22118c3..1455ee0766  HEAD -> sync-from-ra

Error: Josh created a non-roundtrip push! Do NOT merge this into rustc!
```

Reply:
> Very strange. I don't know what to take from your git rev-parse, what do those show?

`HEAD` is the "Merge from downstream" commit, `FETCH_HEAD` is the `josh` change. Or maybe I misunderstood the question?

> You marked the PR as "ready for review" -- did you make more progress here?

No, but I think it's a good start, and the `FETCH_HEAD` thing could conceivably be explained by my disk space problems. And it will be a while before I can try it again (since I have to reinstall and all that).